### ISSUE TITLE
Structure with isolated objects and values that belong to `value:Cloneable` in an isolated expression

### ISSUE THREAD

**Issue by MaryamZi** [ballerina-platform/ballerina-spec#856]
I believe it is currently not possible to make a structure that has both isolated objects and values that belong to `value:Cloneable` directly be part of an isolated expression, which may be required as the initial value of a isolated object's field or when transferring in/out such a structure (in a lock statement that accesses a "protected" variable).

I guess we can still use the spread operator (as suggested in #855) if the values that belong to `value:Cloneable` are subtypes of `readonly`.

```ballerina
type Config record {|
   int[] & readonly a;
   IsolatedClass b;
|};
```

But this would not be possible if the types of the rest of the fields are not subtypes of `readonly`.

```ballerina
type Config record {|
   int[] a;
   IsolatedClass b;
   map<boolean> c;
|};

isolated class Foo {
   private Config c;

   isolated function init(Config con) {
      self.c = con; // ERROR: RHS is not an isolated expression and cannot use `clone`/`cloneReadOnly`.
                    // `self.c = {...con};` can't be used because `a` and `c` are mutable.
   }
}
```

Although we can get this to work by separating out the isolated objects and the values that belong to `value:Cloneable` and only using `clone`/`cloneReadOnly` with the latter, I don't think it will scale with nested records that have isolated objects as field types.

```ballerina
isolated class Foo {
   private Config c;

   isolated function init(Config con) {
      var {b, ...rest} = con;
      self.c = {b, ...rest.clone()};
   }
}
```

Could be related to https://github.com/ballerina-platform/ballerina-spec/issues/784

_Originally posted by @MaryamZi in https://github.com/ballerina-platform/ballerina-spec/issues/855#issuecomment-847628869_

**Comment by MaryamZi** [comment]
> Although we can get this to work by separating out the isolated objects and the values that belong to `value:Cloneable` and only using `clone`/`cloneReadOnly` with the latter, I don't think it will scale with nested records that have isolated objects as field types.

Also won't work with recursive types.

```ballerina
type Client isolated client object {
    remote function get(string... strs) returns anydata|error;
};

type Plugin distinct isolated object {
    isolated function process(Context context) returns error?;
};

type Context record {|
    readonly string path;
    Client 'client;
    Plugin[] plugins;
|};

type Node record {|
    readonly string path;
    Context? context = ();
    table<Node> key(path) children = table [];
|};

isolated class Class {
    private table<Node> key(path) nodes;

    function init(table<Node> key(path) nodeTable) {
        // need to initialize `self.nodes` with values from `nodeTable`
        self.nodes = table []; 
    }
}
```

**Comment by jclark** [comment]
There's a fundamental issue which I think is related to this: #1244.

**Comment by MaryamZi** [comment]
> @MaryamZi Can we close this in favour of #1244?

Yes, I believe the changes proposed in ["Making cloning work with isolated objects" ](https://github.com/ballerina-platform/ballerina-spec/issues/1244#issuecomment-1541355911) addresses this.
